### PR TITLE
Update network.c - Add Bind Address option

### PR DESCRIPTION
**Pull request by elieyal**
This allows the option to add a Bind Address to an IP address instead of the binding to the Interface Name

### CODE REVIEW

File: src/network.c [hunk 1]
```diff
@@ -134,6 +134,7 @@ typedef struct sockent {
   char *node;
   char *service;
   int interface;
+  struct sockaddr_storage *bind_address;
```
Comment (rpv-tomsk) [inline src/network.c]: I think, this should go to `sockent_client` struct.

**Comment by rpv-tomsk** [comment]
Just a note: `https://github.com/redis/hiredis/commit/37d25a392c9b9468e064a67c504939c9c4ea0031`

File: src/network.c [hunk 2]
```diff
@@ -2684,6 +2717,43 @@ static int network_config_set_interface(const oconfig_item_t *ci, /* {{{ */
...
+  int ret;
+  struct addrinfo hint, *res = NULL;
+
+  memset(&hint, '\0', sizeof hint);
+  hint.ai_family = PF_UNSPEC;
+  hint.ai_flags = AI_NUMERICHOST;
+
+  ret = getaddrinfo(addr_text, NULL, &hint, &res);
+  if (ret) {
+    ERROR("Invalid address");
```
Comment (rpv-tomsk) [inline src/network.c]: This needs to be more descriptive and in one line.
Smth like to `ERROR('BindAddress option has invalid address set: %s', gai_strerror())`;

File: src/network.c [hunk 3]
```diff
@@ -1683,6 +1684,36 @@ static int network_set_interface(const sockent_t *se,
...
+    if (bind(se->data.client.fd, (struct sockaddr*)addr, sizeof(*addr)) == -1) {
+      ERROR("network_bind_socket_to_addr: %s", STRERRNO);
+      return -1;
+    }
+  } else if (ai->ai_family == AF_INET6) {
+    struct sockaddr_in6 *addr = (struct sockaddr_in6 *)(se->bind_address);
+    inet_ntop(AF_INET, &(addr->sin6_addr), pbuffer, 64);
+    INFO("binding client socket to ipv6 address: %s", pbuffer);
+    if (bind(se->data.client.fd, (struct sockaddr*)addr, sizeof(*addr)) == -1) {
+      ERROR("network_bind_socket_to_addr: %s", STRERRNO);
```
Comment (rpv-tomsk) [inline src/network.c]: I think `INFO()` messages are excessive. 
And `ERROR()` messages should have prefix `network plugin: ` and be more descriptive.
Common style messages are look like this: `network plugin: Failed to bind client socket (ipvX): %s`, STRERRNO`.



See `network_set_ttl(), network_set_interface()` as reference.

File: src/network.c [hunk 4]
```diff
@@ -2684,6 +2717,43 @@ static int network_config_set_interface(const oconfig_item_t *ci, /* {{{ */
...
+  hint.ai_flags = AI_NUMERICHOST;
+
+  ret = getaddrinfo(addr_text, NULL, &hint, &res);
+  if (ret) {
+    ERROR("Invalid address");
+    ERROR(gai_strerror(ret));
+    return 1;
+  }
+
+  *bind_address = malloc(sizeof(**bind_address));
```
Comment (rpv-tomsk) [inline src/network.c]: Someone can try to put several `BindAddress` options into config.
I think we need take this into account.

Comment (rpv-tomsk) [reply src/network.c]: Allocated memory should be freed at some moment.
Please take a look into `sockent_destroy() / free_sockent_client()`.

**Comment by rpv-tomsk** [comment]
References: 2818

**Comment by rubenk** [comment]
Oops, I missed that we have two pull requests for this and reviewed #2831. @elieyal please address all comments in both PR's and close one of them.

**Pull request closed without merging**